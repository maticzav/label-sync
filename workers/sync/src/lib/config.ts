import { LSCRepository, normalizeColor } from '@labelsync/config'
import { GitHubLabel } from './github'

/**
 * Calculates the difference between labels in the configuration
 * and the labels that are currently in the repository.
 */
export function calculateConfigurationDiff({
  config,
  currentLabels,
}: {
  config: LSCRepository['labels']
  currentLabels: GitHubLabel[]
}): {
  added: GitHubLabel[]
  changed: GitHubLabel[]
  aliased: Required<Pick<GitHubLabel, 'name' | 'old_name'>>[]
  removed: Pick<GitHubLabel, 'name'>[]
} {
  /**
   * Label may fall into four classes:
   *  - it may be new: not on GitHub and no other label
   *    in the configuration references it (e.g. as an old version of the label)
   *  - it may be updated: name remains the same but color
   *    or description have changed
   *  - it may be aliased: one of the labels references a
   *    label that exists on GitHub but is not present in
   *    the configuration anymore
   *    (one label may reference multiple old labels, but multiple
   *      labels can't refernce the same old label)
   *  - it may be removed: no labels alias it and it is
   *    not in the configuration
   *
   * In light of that we diff labels so that:
   *  - added: contains all the labels we should add
   *  - changed: contains all the labels that we should rename,
   *      change description or change color
   *  - aliased: contains all the labels that we should cleverly
   *      reference in the issues
   *  - removed: contains all the labels that we should remove
   */

  const currentLabelsMap = new Map(currentLabels.map((l) => [l.name, l]))

  const added: GitHubLabel[] = []
  const changed: GitHubLabel[] = []
  const aliased: Required<Pick<GitHubLabel, 'name' | 'old_name'>>[] = []
  const removed: Pick<GitHubLabel, 'name'>[] = []

  for (const label of Object.keys(config)) {
    // This algorithm assumes each label is only referenced once.
    // You may not reference the same label in two different alias.
    const hydratedLabel = hydrateLabel(label)

    const isExistingLabel = hydratedLabel.old_name !== undefined

    const aliaii: string[] = (config[label].alias ?? []).filter((aliasName) => currentLabelsMap.has(aliasName))
    const isLabelAlias = aliaii.length > 0

    const hasLabelChanged = currentLabels.some((cLabel) => {
      if (cLabel.name !== hydratedLabel.name) {
        return false
      }

      // Description is optional. It has changed if:
      //  - it wasn't defined and now is
      //  - it was defined and has changed
      const hasDescriptionChanged = (cLabel.description ?? '') !== (hydratedLabel.description ?? '')

      // Color of a label is always defined. We check whether it has changed.
      const hasColorChanged = normalizeColor(cLabel.color) !== normalizeColor(hydratedLabel.color)

      return hasDescriptionChanged || hasColorChanged
    })

    // New label:
    //  - doesn't exist yet
    //  - isn't an alias for another label
    if (!isExistingLabel && !isLabelAlias) {
      added.push(hydratedLabel)
      continue
    }

    // Updated labels:
    //  - either color or description has changed.
    if (isExistingLabel && hasLabelChanged && !isLabelAlias) {
      changed.push(hydratedLabel)
      continue
    }

    // Aliased labels:
    //  - add an existing label to changed labels
    //  - this one shouldn't be removed afterwards since we
    //    are renaming it.

    // Find the first old label of the label that is currently examined
    // and use it as the source that should be changed.
    const origin = aliaii.find((alias) => currentLabelsMap.has(alias))

    for (const alias of aliaii) {
      if (alias === origin && !isExistingLabel) {
        // Change the "original" label to the name of the new one.
        changed.push({ ...hydratedLabel, old_name: alias })
      } else {
        // Alias all other labels to the new one.
        aliased.push({ ...hydratedLabel, old_name: alias })
      }
    }
  }

  for (const label of currentLabels) {
    const isLabelConfigured = config.hasOwnProperty(label.name)
    const isLabelRenamed = changed.find(({ old_name }) => old_name === label.name)

    // Mark label as removed if it weren't renamed (changed)
    // as part of the aliasing and isn't configured anymore.
    //
    // NOTE: This also catches aliased labels that won't be renamed.
    if (!isLabelConfigured && !isLabelRenamed) {
      removed.push({ name: label.name })
    }
  }

  return {
    added,
    changed,
    removed,
    aliased,
  }

  /**
   * Returns information for the label with a given name from the
   * currently active labels map.
   */
  function hydrateLabel(name: string): GitHubLabel {
    const oldLabel = currentLabelsMap.get(name)

    return {
      old_name: oldLabel?.name,
      name,
      //
      old_description: oldLabel?.description,
      description: config[name]!.description,
      //
      old_color: oldLabel?.color,
      color: config[name]!.color,
    }
  }
}
