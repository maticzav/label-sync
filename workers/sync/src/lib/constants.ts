import ml from 'multilines'
import os from 'os'

export const messages = {
  'insufficient.permissions': (missingRepos: string[]) => ml`
	| # Insufficient permissions
	|
	| Hi there,
	| Thank you for installing LabelSync. We have noticed that your configuration stretches beyond repositories we can access. 
	| We think you forgot to allow access to certain repositories. Please update your installation. 
	|
	| _Missing repositories:_
	| ${missingRepos.map((missing) => ` * ${missing}`).join(os.EOL)}
	|
	| Best,
	| LabelSync Team
	`,
  'onboarding.error': (error: string) => ml`
	| # Welcome to LabelSync!
	|
	| Hi there,
	| Thank you for using LabelSync. We hope you enjoyed the experience so far.
	| It seems like there are some problems with your configuration.
	| Our parser reported that:
	|
	| ${error}
	|
	| Let us know if we can help you with the configuration by sending us an email to support@label-sync.com. 
	| We'll try to get back to you as quickly as possible.
	|
	| Best,
	| LabelSync Team
	`,
}
