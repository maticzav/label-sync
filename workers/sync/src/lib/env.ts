type Config = {
  prod: boolean
  ghAppId: string
  ghPrivateKey: string

  redisUrl: string
}

// Environments

const base = {}

const prod = {
  prod: true,
  ghAppId: process.env.GH_APP_ID!,
  ghPrivateKey: process.env.GH_PRIVATE_KEY!,

  redisUrl: process.env.REDIS_URL!,
}

const dev = {
  prod: false,
  ghAppId: '',
  ghPrivateKey: '',

  redisUrl: 'redis://localhost:6379',
}

const enviroment = process.env.NODE_ENV

/**
 * Configuration credentials for the worker instance.
 */
export const config: Config = Object.assign(base, enviroment === 'production' ? prod : dev)
