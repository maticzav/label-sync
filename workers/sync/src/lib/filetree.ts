import * as path from 'path'

import { Dict } from '../data/dict'

/**
 * Represents a Github file and folder structure where
 * files are represented as UTF-8 encoded strings keyed by the their relative
 * path to the repository root.
 */
export type FileTree = { [path: string]: string }

/**
 * Represents a file and folder structure where
 * files are represented as UTF-8 encoded strings keyed by the their relative
 * path to the root directory.
 */
export namespace FileTree {
  export type Type = { [path: string]: string }

  /**
   * Returns the files that are not nested.
   */
  export function getRootFiles(tree: FileTree): Dict<string> {
    const files = Object.keys(tree)
      .filter(isFileInRootFolder)
      .map((name) => [name, tree[name]])

    return Object.fromEntries(files)
  }

  /**
   * Returns a dictionary of remaining subtrees.
   */
  export function getSubtrees(tree: FileTree): Dict<FileTree> {
    return Object.keys(tree)
      .filter((file) => !isFileInRootFolder(file))
      .reduce<Dict<FileTree>>((acc, filepath) => {
        const [subTree, newFilepath] = shiftPath(filepath)
        if (!acc.hasOwnProperty(subTree)) {
          acc[subTree] = {}
        }
        acc[subTree][newFilepath] = tree[filepath]
        return acc
      }, {})
  }

  /**
   * Shifts path by one and returns the shifted part as first element in tuple
   *  and remaining part as the second.
   */
  function shiftPath(filepath: string): [string, string] {
    const [dir, ...dirs] = filepath.split('/').filter(Boolean)
    return [dir, dirs.join('/')]
  }

  /**
   * Determines whether a path references a direct file
   * or a file in the nested folder.
   *
   * "/src/index.ts" -> false
   * "/index.ts" -> true
   * "index.ts" -> true
   */
  function isFileInRootFolder(filePath: string): boolean {
    return ['.', '/'].includes(path.dirname(filePath))
  }
}
