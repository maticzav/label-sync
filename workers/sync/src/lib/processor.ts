import { ITaskQueue } from '@labelsync/queues'
import pino from 'pino'

import { IGitHubEndpoints } from './github'

/**
 * A blueprint specification for a task processor that all processors
 * should extend. It provides base functionality for the processor.
 */
export class Processor<T> {
  protected installation: { id: number }

  /**
   * Available GitHub methods to communicate with the API.
   */
  protected endpoints: IGitHubEndpoints

  /**
   * A general logger that may be used to log significant events
   * that occur during execution.
   */
  protected log: pino.Logger

  /**
   * A queue that lets you push new tasks to the queue.
   */
  protected queue: Pick<ITaskQueue, 'push'>

  constructor(
    installation: { id: number },
    queue: Pick<ITaskQueue, 'push'>,
    endpoints: IGitHubEndpoints,
    logger: pino.Logger,
  ) {
    this.endpoints = endpoints
    this.installation = installation
    this.log = logger
    this.queue = queue
  }

  /**
   * Performs the current task and returns new tasks to be executed.
   */
  perform(data: T): Promise<void> {
    throw new Error(`Missing perform implementation in processor.`)
  }
}
