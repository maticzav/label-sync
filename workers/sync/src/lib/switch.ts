/**
 * Makes a type check that is only valid when all cases of a switch
 * statement have been convered.
 */
export class ExhaustiveSwitchCheck extends Error {
  constructor(val: never) {
    super(`Unreachable case: ${JSON.stringify(val)}`)
  }
}
