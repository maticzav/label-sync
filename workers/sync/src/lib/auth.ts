import { Octokit } from '@octokit/core'
import { createAppAuth, createOAuthUserAuth } from '@octokit/auth-app'

import { config } from './env'

/**
 * Returns an instance of Octokit that is authenticated as the
 * given installation of the app and automatically refreshes the token.
 *
 * https://github.com/octokit/auth-app.js#usage-with-octokit
 */
export const getOctokitForInstallation = (installationId: number): Octokit => {
  const octokit = new Octokit({
    authStrategy: createAppAuth,
    auth: { appId: config.ghAppId, privateKey: config.ghPrivateKey, installationId },
  })

  return octokit
}
