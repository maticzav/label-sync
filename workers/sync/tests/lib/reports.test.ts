import { generateHumanReadablePRReport, generateHumanReadableCommitReport } from '../../src/lib/reports'
import * as reports from '../__fixtures__/reports'

describe('reports', () => {
  for (const report of [reports.strict, reports.failure, reports.nonstrict, reports.unchanged, reports.unconfigured]) {
    test('correcly generates a PR report', () => {
      const message = generateHumanReadablePRReport([report])
      expect(message).toMatchSnapshot()
    })

    test('correctly generates a commit report', () => {
      const message = generateHumanReadableCommitReport([report])
      expect(message).toMatchSnapshot()
    })
  }

  test('correctly combines multiple reports', () => {
    const report = generateHumanReadablePRReport([
      reports.strict,
      reports.nonstrict,
      reports.failure,
      reports.unchanged,
    ])

    expect(report).toMatchSnapshot()
  })
})
