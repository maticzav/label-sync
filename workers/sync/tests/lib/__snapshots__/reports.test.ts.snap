// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reports correcly generates a PR report 1`] = `
"## :label: Here's what's going to change when you merge

#### \`success\`

<details>
  <summary>:sparkles: 2 new labels.</summary>

- ![addition/one](https://img.shields.io/static/v1?color=%23ee263c&label=&message=addition%2Fone \\"addition/one\\")
- ![addition/two](https://img.shields.io/static/v1?color=%23ee263c&label=&message=addition%2Ftwo \\"addition/two\\")

</details>

<details>
  <summary>:lipstick: 3 changed labels.</summary>

- ![update/ena (name changed)](https://img.shields.io/static/v1?color=%23FDE8E8&label=&message=update%2Fena%20%28name%20changed%29 \\"update/ena (name changed)\\")
- ![update/dve (name and color changed)](https://img.shields.io/static/v1?color=%23FDF6B2&label=&message=update%2Fdve%20%28name%20and%20color%20changed%29 \\"update/dve (name and color changed)\\")
- ![update/tri (name, description and color changed)](https://img.shields.io/static/v1?color=%23E1EFFE&label=&message=update%2Ftri%20%28name%2C%20description%20and%20color%20changed%29 \\"update/tri (name, description and color changed)\\")

</details>

<details>
  <summary>:performing_arts: 2 aliases.</summary>

- ![label/alias (name changed)](https://img.shields.io/static/v1?color=%23ee263c&label=&message=label%2Falias%20%28name%20changed%29 \\"label/alias (name changed)\\")
- ![label/alias (name changed)](https://img.shields.io/static/v1?color=%23ee263c&label=&message=label%2Falias%20%28name%20changed%29 \\"label/alias (name changed)\\")

</details>

<details>
  <summary>:coffin: 2 removed labels.</summary>

- ![removal/one](https://img.shields.io/static/v1?color=%23aa123c&label=&message=removal%2Fone \\"removal/one\\")
- ![removal/two](https://img.shields.io/static/v1?color=%23aa123c&label=&message=removal%2Ftwo \\"removal/two\\")

</details>

---

### You haven't made any changes in these repositories

No unchanged repositories.
"
`;

exports[`reports correcly generates a PR report 2`] = `
"## :label: Here's what's going to change when you merge

#### \`failure\`

> ❗️ LabelSync encountered a problem syncing this repository.

Couldn't make a diff of labels.

---

### You haven't made any changes in these repositories

No unchanged repositories.
"
`;

exports[`reports correcly generates a PR report 3`] = `
"## :label: Here's what's going to change when you merge

#### \`unconfigured-nonstrict\`

<details>
  <summary>:mailbox_with_mail: 2 unconfigured labels.</summary>

- ![unconfigured-one](https://img.shields.io/static/v1?color=%23ee263c&label=&message=unconfigured-one \\"unconfigured-one\\")
- ![unconfigured-two](https://img.shields.io/static/v1?color=%23ee263c&label=&message=unconfigured-two \\"unconfigured-two\\")

</details>

---

### You haven't made any changes in these repositories

No unchanged repositories.
"
`;

exports[`reports correcly generates a PR report 4`] = `
"## :label: Here's what's going to change when you merge

---

### You haven't made any changes in these repositories

<details>
  <summary>1 repositories.</summary>

- \`unchanged\`

</details>
"
`;

exports[`reports correcly generates a PR report 5`] = `
"## :label: Here's what's going to change when you merge

#### \`unconfigured\`

<details>
  <summary>:sparkles: 1 new label.</summary>

- ![addition](https://img.shields.io/static/v1?color=%23ee263c&label=&message=addition \\"addition\\")

</details>

> You have no unconfigured labels - you could make this repository \`strict\`.

---

### You haven't made any changes in these repositories

No unchanged repositories.
"
`;

exports[`reports correctly combines multiple reports 1`] = `
"## :label: Here's what's going to change when you merge

#### \`failure\`

> ❗️ LabelSync encountered a problem syncing this repository.

Couldn't make a diff of labels.

---

#### \`success\`

<details>
  <summary>:sparkles: 2 new labels.</summary>

- ![addition/one](https://img.shields.io/static/v1?color=%23ee263c&label=&message=addition%2Fone \\"addition/one\\")
- ![addition/two](https://img.shields.io/static/v1?color=%23ee263c&label=&message=addition%2Ftwo \\"addition/two\\")

</details>

<details>
  <summary>:lipstick: 3 changed labels.</summary>

- ![update/ena (name changed)](https://img.shields.io/static/v1?color=%23FDE8E8&label=&message=update%2Fena%20%28name%20changed%29 \\"update/ena (name changed)\\")
- ![update/dve (name and color changed)](https://img.shields.io/static/v1?color=%23FDF6B2&label=&message=update%2Fdve%20%28name%20and%20color%20changed%29 \\"update/dve (name and color changed)\\")
- ![update/tri (name, description and color changed)](https://img.shields.io/static/v1?color=%23E1EFFE&label=&message=update%2Ftri%20%28name%2C%20description%20and%20color%20changed%29 \\"update/tri (name, description and color changed)\\")

</details>

<details>
  <summary>:performing_arts: 2 aliases.</summary>

- ![label/alias (name changed)](https://img.shields.io/static/v1?color=%23ee263c&label=&message=label%2Falias%20%28name%20changed%29 \\"label/alias (name changed)\\")
- ![label/alias (name changed)](https://img.shields.io/static/v1?color=%23ee263c&label=&message=label%2Falias%20%28name%20changed%29 \\"label/alias (name changed)\\")

</details>

<details>
  <summary>:coffin: 2 removed labels.</summary>

- ![removal/one](https://img.shields.io/static/v1?color=%23aa123c&label=&message=removal%2Fone \\"removal/one\\")
- ![removal/two](https://img.shields.io/static/v1?color=%23aa123c&label=&message=removal%2Ftwo \\"removal/two\\")

</details>

---

#### \`unconfigured-nonstrict\`

<details>
  <summary>:mailbox_with_mail: 2 unconfigured labels.</summary>

- ![unconfigured-one](https://img.shields.io/static/v1?color=%23ee263c&label=&message=unconfigured-one \\"unconfigured-one\\")
- ![unconfigured-two](https://img.shields.io/static/v1?color=%23ee263c&label=&message=unconfigured-two \\"unconfigured-two\\")

</details>

---

### You haven't made any changes in these repositories

<details>
  <summary>1 repositories.</summary>

- \`unchanged\`

</details>
"
`;

exports[`reports correctly generates a commit report 1`] = `
"## :label: Here's what has changed.

#### \`success\`

<details>
  <summary>:sparkles: 2 new labels.</summary>

- ![addition/one](https://img.shields.io/static/v1?color=%23ee263c&label=&message=addition%2Fone \\"addition/one\\")
- ![addition/two](https://img.shields.io/static/v1?color=%23ee263c&label=&message=addition%2Ftwo \\"addition/two\\")

</details>

<details>
  <summary>:lipstick: 3 changed labels.</summary>

- ![update/ena (name changed)](https://img.shields.io/static/v1?color=%23FDE8E8&label=&message=update%2Fena%20%28name%20changed%29 \\"update/ena (name changed)\\")
- ![update/dve (name and color changed)](https://img.shields.io/static/v1?color=%23FDF6B2&label=&message=update%2Fdve%20%28name%20and%20color%20changed%29 \\"update/dve (name and color changed)\\")
- ![update/tri (name, description and color changed)](https://img.shields.io/static/v1?color=%23E1EFFE&label=&message=update%2Ftri%20%28name%2C%20description%20and%20color%20changed%29 \\"update/tri (name, description and color changed)\\")

</details>

<details>
  <summary>:performing_arts: 2 aliases.</summary>

- ![label/alias (name changed)](https://img.shields.io/static/v1?color=%23ee263c&label=&message=label%2Falias%20%28name%20changed%29 \\"label/alias (name changed)\\")
- ![label/alias (name changed)](https://img.shields.io/static/v1?color=%23ee263c&label=&message=label%2Falias%20%28name%20changed%29 \\"label/alias (name changed)\\")

</details>

<details>
  <summary>:coffin: 2 removed labels.</summary>

- ![removal/one](https://img.shields.io/static/v1?color=%23aa123c&label=&message=removal%2Fone \\"removal/one\\")
- ![removal/two](https://img.shields.io/static/v1?color=%23aa123c&label=&message=removal%2Ftwo \\"removal/two\\")

</details>
"
`;

exports[`reports correctly generates a commit report 2`] = `
"## :label: Here's what has changed.

#### \`failure\`

> ❗️ LabelSync encountered a problem syncing this repository.

Couldn't make a diff of labels.
"
`;

exports[`reports correctly generates a commit report 3`] = `
"## :label: Here's what has changed.
"
`;

exports[`reports correctly generates a commit report 4`] = `
"## :label: Here's what has changed.
"
`;

exports[`reports correctly generates a commit report 5`] = `
"## :label: Here's what has changed.

#### \`unconfigured\`

<details>
  <summary>:sparkles: 1 new label.</summary>

- ![addition](https://img.shields.io/static/v1?color=%23ee263c&label=&message=addition \\"addition\\")

</details>

> You have no unconfigured labels - you could make this repository \`strict\`.
"
`;
